flo-runner/src/main/java/com/spotify/flo/context/ForkingExecutor.java

```diff
@@ -0,0 +1,378 @@
+      processBuilder.command().add(closureFile.toString());
+      processBuilder.command().add(resultFile.toString());
+      processBuilder.command().add(errorFile.toString());
+
+      processBuilder.environment().putAll(environment);
+
+      log.debug("Starting subprocess: environment={}, command={}, directory={}",
+          processBuilder.environment(), processBuilder.command(), processBuilder.directory());
+      try {
+        process = processBuilder.start();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+
+      // Copy std{err,out} line by line to avoid interleaving and corrupting line contents.
+      executor.submit(() -> copyLines(process.getInputStream(), System.out));
+      executor.submit(() -> copyLines(process.getErrorStream(), System.err));
+    }
+
+    T waitFor() {
+      if (process == null) {
+        throw new IllegalStateException();
+      }
+      log.debug("Waiting for subprocess exit");
+      final int exitValue;
+      try {
+        exitValue = process.waitFor();
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new RuntimeException(e);
+      } finally {
+        process.destroyForcibly();
+      }
+
+      log.debug("Subprocess exited: " + exitValue);
+      if (exitValue != 0) {
+        throw new RuntimeException("Subprocess failed: " + process.exitValue());
+      }
+
+      if (Files.exists(errorFile)) {
+        // Failed
+        log.debug("Subprocess exited with error file");
+        final Throwable error;
+        try {
+          error = PersistingContext.deserialize(errorFile);
+        } catch (Exception e) {
+          throw new RuntimeException("Failed to deserialize error", e);
+        }
+        if (error instanceof RuntimeException) {
+          throw (RuntimeException) error;
+        } else {
+          throw new RuntimeException(error);
+        }
+      } else {
+        // Success
+        log.debug("Subprocess exited with result file");
+        final T result;
+        try {
+          result = PersistingContext.deserialize(resultFile);
+        } catch (Exception e) {
+          throw new RuntimeException("Failed to deserialize result", e);
+        }
+        return result;
+      }
+    }
+
+    @Override
+    public void close() {
+      if (process != null) {
+        process.destroyForcibly();
+        process = null;
+      }
+      executor.shutdown();
+      tryDeleteDir(tempdir);
+    }
+  }
+
+  private static void tryDeleteDir(Path path) {
+    try {
+      deleteDir(path);
+    } catch (IOException e) {
+      log.warn("Failed to delete directory: {}", path, e);
+    }
+  }
+
+  private static void deleteDir(Path path) throws IOException {
+    try {
+      Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
+        @Override
+        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+          try {
+            Files.delete(file);
```

Comment:
What would cause concurrent deletion, or just to be extra cautious?

Reply:
`close()` is called both [here](https://github.com/spotify/flo/pull/84/files#diff-2df9957d7823bf437f9f3c51ceff75a6R87) and [here](https://github.com/spotify/flo/pull/84/files#diff-2df9957d7823bf437f9f3c51ceff75a6R82). A bit extra cautios code I guess.